// ngx-datatable-footer/esm2015/lib/datatable-footer/datatable-footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, TemplateRef } from '@angular/core';
import { FooterConfigurationService } from '../services/configuration.service';
export class DatatableFooterComponent {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.config = config;
        this.pagerLeftArrowIcon = 'fa fa-angle-left';
        this.pagerRightArrowIcon = 'fa fa-angle-right';
        this.pagerPreviousIcon = 'fa fa-step-backward';
        this.pagerNextIcon = 'fa fa-step-forward';
        this.pageSizes = [10, 20, 30, 40, 50];
        this.totalNumber = true;
        this.linage = true;
        this.displayBothEnds = true;
        this.pageNumber = true;
        this.pageJumpTime = 1000;
        this.pageChange = new EventEmitter();
        this.maxPage = new EventEmitter();
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set template(val) {
        this._template = val;
        this.update();
    }
    /**
     * @return {?}
     */
    get template() {
        return this._template;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set datatable(val) {
        this._datatable = val;
        this.update();
    }
    /**
     * @return {?}
     */
    get datatable() {
        return this._datatable;
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    update() {
        if (this._datatable &&
            this._template &&
            (!this._datatable.footer || !this._datatable.footer.template)) {
            if (this._datatable.footer) {
                this._datatable.footer.template = this._template;
            }
            else {
                this._datatable.footer = {
                    footerHeight: this._datatable.footerHeight,
                    totalMessage: this._datatable.messages.totalMessage,
                    selectedMessage: this._datatable.messages.selectedMessage,
                    pagerLeftArrowIcon: '',
                    pagerRightArrowIcon: '',
                    pagerPreviousIcon: '',
                    pagerNextIcon: '',
                    template: this._template
                };
            }
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onFooterPage(event) {
        if (typeof event.page === 'number') {
            this.datatable.onFooterPage(event);
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    emitPage(event) {
        if (typeof event.page === 'number') {
            this.pageChange.emit(event);
        }
    }
    /**
     * @param {?} page
     * @return {?}
     */
    emitMaxPage(page) {
        this.maxPage.emit(page);
    }
}
DatatableFooterComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-datatable-footer',
                template: ` <ng-template
    ngx-datatable-footer-template
    let-rowCount="rowCount"
    let-pageSize="pageSize"
    let-selectedCount="selectedCount"
    let-curPage="curPage"
    let-offset="offset"
  >
    <div
      class="page-count"
      *ngIf="rowCount > 0 && totalNumber"
      [innerHTML]="
        config.labels.current +
        ((offset || 0) * pageSize + 1) +
        ' - ' +
        ((offset + 1) * pageSize > rowCount
          ? rowCount
          : (offset + 1 || 1) * pageSize) +
        ' / ' +
        rowCount +
        config.labels.line
      "
    ></div>
    <app-datatable-pager
      *ngIf="rowCount > 0"
      [linage]="linage"
      [displayBothEnds]="displayBothEnds"
      [pageNumber]="pageNumber"
      [pageSizes]="pageSizes"
      [pagerPreviousIcon]="pagerPreviousIcon"
      [pagerNextIcon]="pagerNextIcon"
      [pagerLeftArrowIcon]="pagerLeftArrowIcon"
      [pagerRightArrowIcon]="pagerRightArrowIcon"
      [page]="curPage"
      [size]="pageSize"
      [pageJumpTime]="pageJumpTime"
      [datatable]="datatable"
      [count]="rowCount"
      [labels]="config.labels"
      (change)="onFooterPage($event); emitPage($event)"
      (pageChange)="emitPage($event)"
      (maxPage)="emitMaxPage($event)"
      class="datatable-pager"
    >
    </app-datatable-pager>
  </ng-template>`
            }] }
];
/** @nocollapse */
DatatableFooterComponent.ctorParameters = () => [
    { type: FooterConfigurationService }
];
DatatableFooterComponent.propDecorators = {
    template: [{ type: ViewChild, args: [TemplateRef,] }],
    datatable: [{ type: Input }],
    pagerLeftArrowIcon: [{ type: Input }],
    pagerRightArrowIcon: [{ type: Input }],
    pagerPreviousIcon: [{ type: Input }],
    pagerNextIcon: [{ type: Input }],
    pageSizes: [{ type: Input }],
    totalNumber: [{ type: Input }],
    linage: [{ type: Input }],
    displayBothEnds: [{ type: Input }],
    pageNumber: [{ type: Input }],
    page: [{ type: Input }],
    pageJumpTime: [{ type: Input }],
    pageChange: [{ type: Output }],
    maxPage: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DatatableFooterComponent.prototype.pagerLeftArrowIcon;
    /** @type {?} */
    DatatableFooterComponent.prototype.pagerRightArrowIcon;
    /** @type {?} */
    DatatableFooterComponent.prototype.pagerPreviousIcon;
    /** @type {?} */
    DatatableFooterComponent.prototype.pagerNextIcon;
    /** @type {?} */
    DatatableFooterComponent.prototype.pageSizes;
    /** @type {?} */
    DatatableFooterComponent.prototype.totalNumber;
    /** @type {?} */
    DatatableFooterComponent.prototype.linage;
    /** @type {?} */
    DatatableFooterComponent.prototype.displayBothEnds;
    /** @type {?} */
    DatatableFooterComponent.prototype.pageNumber;
    /** @type {?} */
    DatatableFooterComponent.prototype.page;
    /** @type {?} */
    DatatableFooterComponent.prototype.pageJumpTime;
    /** @type {?} */
    DatatableFooterComponent.prototype.pageChange;
    /** @type {?} */
    DatatableFooterComponent.prototype.maxPage;
    /**
     * @type {?}
     * @private
     */
    DatatableFooterComponent.prototype._template;
    /**
     * @type {?}
     * @private
     */
    DatatableFooterComponent.prototype._datatable;
    /** @type {?} */
    DatatableFooterComponent.prototype.config;
}
