// ng-block-ui/fesm2015/ng-block-ui-router.js
import { __decorate } from 'tslib';
import { Injectable, NgModule } from '@angular/core';
import { BLOCKUI_DEFAULT, BlockUIService, BlockUIModule } from 'ng-block-ui';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ng-block-ui';
let BlockUIPreventNavigation = class BlockUIPreventNavigation {
    constructor(blockUIService) {
        this.blockUIService = blockUIService;
    }
    canActivate() {
        return !this.blockUIService.isActive(BLOCKUI_DEFAULT);
    }
    canActivateChild() {
        return !this.blockUIService.isActive(BLOCKUI_DEFAULT);
    }
};
BlockUIPreventNavigation.ɵfac = function BlockUIPreventNavigation_Factory(t) { return new (t || BlockUIPreventNavigation)(ɵngcc0.ɵɵinject(ɵngcc1.BlockUIService)); };
BlockUIPreventNavigation.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: BlockUIPreventNavigation, factory: function (t) { return BlockUIPreventNavigation.ɵfac(t); } });
BlockUIPreventNavigation.ctorParameters = () => [
    { type: BlockUIService }
];

var BlockUIRouterModule_1;
let BlockUIRouterModule = BlockUIRouterModule_1 = class BlockUIRouterModule {
    static forRoot() {
        return {
            ngModule: BlockUIRouterModule_1,
            providers: [
                BlockUIPreventNavigation
            ]
        };
    }
};
BlockUIRouterModule.ɵfac = function BlockUIRouterModule_Factory(t) { return new (t || BlockUIRouterModule)(); };
BlockUIRouterModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BlockUIRouterModule });
BlockUIRouterModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            BlockUIModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BlockUIPreventNavigation, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.BlockUIService }]; }, null); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BlockUIRouterModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BlockUIModule
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BlockUIRouterModule, { imports: function () { return [BlockUIModule]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { BlockUIPreventNavigation, BlockUIRouterModule };
