// ng-sidebar/lib/close.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var sidebar_component_1 = require("./sidebar.component");
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./sidebar.component');
var CloseSidebar = /** @class */ (function () {
    function CloseSidebar(_sidebar) {
        this._sidebar = _sidebar;
    }
    /** @internal */
    CloseSidebar.prototype._onClick = function () {
        if (this._sidebar) {
            this._sidebar.close();
        }
    };
    /** @nocollapse */
    CloseSidebar.ctorParameters = function () { return [
        { type: sidebar_component_1.Sidebar }
    ]; };
CloseSidebar.ɵfac = function CloseSidebar_Factory(t) { return new (t || CloseSidebar)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Sidebar)); };
CloseSidebar.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: CloseSidebar, selectors: [["", "closeSidebar", ""]], hostBindings: function CloseSidebar_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function CloseSidebar_click_HostBindingHandler() { return ctx._onClick(); });
    } } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CloseSidebar, [{
        type: core_1.Directive,
        args: [{
                selector: '[closeSidebar]',
                host: {
                    '(click)': '_onClick()'
                }
            }]
    }], function () { return [{ type: ɵngcc1.Sidebar }]; }, null); })();
    return CloseSidebar;
}());
exports.CloseSidebar = CloseSidebar;
