// ng-sidebar/lib/sidebar.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var sidebar_container_component_1 = require("./sidebar-container.component");
var sidebar_component_1 = require("./sidebar.component");
var close_directive_1 = require("./close.directive");
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./sidebar-container.component');
var ɵngcc2 = require('./sidebar.component');
var ɵngcc3 = require('./close.directive');
var ɵngcc4 = require('@angular/common');
var SidebarModule = /** @class */ (function () {
    function SidebarModule() {
    }
    SidebarModule.forRoot = function () {
        return {
            ngModule: SidebarModule
        };
    };
SidebarModule.ɵfac = function SidebarModule_Factory(t) { return new (t || SidebarModule)(); };
SidebarModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SidebarModule });
SidebarModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[common_1.CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SidebarModule, [{
        type: core_1.NgModule,
        args: [{
                declarations: [sidebar_container_component_1.SidebarContainer, sidebar_component_1.Sidebar, close_directive_1.CloseSidebar],
                imports: [common_1.CommonModule],
                exports: [sidebar_container_component_1.SidebarContainer, sidebar_component_1.Sidebar, close_directive_1.CloseSidebar]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SidebarModule, { declarations: [ɵngcc1.SidebarContainer, ɵngcc2.Sidebar, ɵngcc3.CloseSidebar], imports: [ɵngcc4.CommonModule], exports: [ɵngcc1.SidebarContainer, ɵngcc2.Sidebar, ɵngcc3.CloseSidebar] }); })();
    return SidebarModule;
}());
exports.SidebarModule = SidebarModule;
