// ngx-datatable-footer/esm5/lib/ngx-datatable-footer.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FooterConfigurationService } from './services/configuration.service';
import { DatatablePagerComponent } from './datatable-pager/datatable-pager.component';
import { DatatableFooterComponent } from './datatable-footer/datatable-footer.component';
var NgxDatatableFooterModule = /** @class */ (function () {
    function NgxDatatableFooterModule() {
    }
    /**
     * @return {?}
     */
    NgxDatatableFooterModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: NgxDatatableFooterModule,
        };
    };
    NgxDatatableFooterModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule
                    ],
                    declarations: [
                        DatatablePagerComponent,
                        DatatableFooterComponent
                    ],
                    exports: [
                        DatatablePagerComponent,
                        DatatableFooterComponent
                    ],
                    providers: [
                        FooterConfigurationService
                    ]
                },] }
    ];
    return NgxDatatableFooterModule;
}());
export { NgxDatatableFooterModule };
