// ngx-datatable-footer/esm2015/lib/services/configuration.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
export class FooterConfigurationService {
    /**
     * @param {?} translate
     */
    constructor(translate) {
        this.translate = translate;
        this.defaultLabels = {
            current: '当前 ',
            line: '行',
            page: '页',
            jumpTo: '跳转到',
        };
        this.update();
        this.translate.onLangChange.subscribe((lang) => {
            setTimeout(() => {
                this.update();
            }, 0);
        });
    }
    /**
     * @return {?}
     */
    update() {
        this._language = localStorage.getItem("lang");
        if (this._language === "zh-CN" || !this._language) {
            this.labels = this.defaultLabels;
            // this.labels = Object.assign({}, this.labels, options.labels);
        }
        else {
            if (this._language) {
                if (this._language) {
                    this.translate.get("ngxFooterSettings.labels").subscribe((translations) => {
                        this.labels = translations;
                    });
                }
            }
        }
    }
}
FooterConfigurationService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FooterConfigurationService.ctorParameters = () => [
    { type: TranslateService }
];
if (false) {
    /** @type {?} */
    FooterConfigurationService.prototype.labels;
    /** @type {?} */
    FooterConfigurationService.prototype.defaultLabels;
    /**
     * @type {?}
     * @private
     */
    FooterConfigurationService.prototype._language;
    /**
     * @type {?}
     * @private
     */
    FooterConfigurationService.prototype.translate;
}
