// ng-block-ui/bundles/ng-block-ui.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('ng-block-ui', ['exports', '@angular/core', '@angular/common', 'rxjs'], factory) :
    (global = global || self, factory(global['ng-block-ui'] = {}, global.ng.core, global.ng.common, global.rxjs));
}(this, (function (exports, core, common, rxjs) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var BlockUIActions = /** @class */ (function () {
        function BlockUIActions() {
        }
        BlockUIActions.START = 'start';
        BlockUIActions.STOP = 'stop';
        BlockUIActions.UPDATE = 'update';
        BlockUIActions.RESET = 'reset';
        BlockUIActions.RESET_GLOBAL = 'reset_global';
        BlockUIActions.UNSUBSCRIBE = 'unsubscribe';
        return BlockUIActions;
    }());

    var BlockUIDefaultName = 'block-ui-main';

    var BlockUIInstanceService = /** @class */ (function () {
        function BlockUIInstanceService() {
            this.blockUISettings = {};
            this.blockUIInstances = {};
            this.blockUISubject = new rxjs.ReplaySubject(1);
            this.blockUIObservable = this.blockUISubject.asObservable();
            this.blockUIObservable.subscribe(this.blockUIMiddleware.bind(this));
        }
        BlockUIInstanceService.prototype.getSettings = function () {
            return this.blockUISettings;
        };
        BlockUIInstanceService.prototype.updateSettings = function (settings) {
            if (settings === void 0) { settings = {}; }
            this.blockUISettings = __assign(__assign({}, this.blockUISettings), settings);
        };
        BlockUIInstanceService.prototype.decorate = function (name) {
            if (name === void 0) { name = BlockUIDefaultName; }
            var blockUI = {
                name: name,
                isActive: false,
                blockCount: 0,
                start: this.dispatch(this.blockUISubject, BlockUIActions.START, name),
                update: this.dispatch(this.blockUISubject, BlockUIActions.UPDATE, name),
                stop: this.dispatch(this.blockUISubject, BlockUIActions.STOP, name),
                reset: this.dispatch(this.blockUISubject, BlockUIActions.RESET, name),
                resetGlobal: this.dispatch(this.blockUISubject, BlockUIActions.RESET_GLOBAL, name),
                unsubscribe: this.dispatch(this.blockUISubject, BlockUIActions.UNSUBSCRIBE, name)
            };
            this.blockUIInstances[name] = this.blockUIInstances[name] || blockUI;
            return blockUI;
        };
        BlockUIInstanceService.prototype.observe = function () {
            return this.blockUIObservable;
        };
        BlockUIInstanceService.prototype.clearInstance = function (instanceName) {
            this.dispatch(this.blockUISubject, BlockUIActions.RESET, instanceName);
        };
        BlockUIInstanceService.prototype.blockUIMiddleware = function (_a) {
            var action = _a.action, name = _a.name;
            var isActive = null;
            switch (action) {
                case (BlockUIActions.START):
                    isActive = true;
                    break;
                case (BlockUIActions.STOP):
                case (BlockUIActions.RESET):
                    isActive = false;
                    break;
            }
            if (isActive !== null) {
                this.blockUIInstances[name].isActive = isActive;
            }
        };
        BlockUIInstanceService.prototype.dispatch = function (subject, action, name) {
            if (name === void 0) { name = BlockUIDefaultName; }
            return function (message) {
                subject.next({
                    name: name,
                    action: action,
                    message: message
                });
            };
        };
        BlockUIInstanceService = __decorate([
            core.Injectable()
        ], BlockUIInstanceService);
        return BlockUIInstanceService;
    }());

    var BlockUIComponent = /** @class */ (function () {
        function BlockUIComponent(blockUI) {
            this.blockUI = blockUI;
        }
        BlockUIComponent.prototype.ngOnInit = function () {
            this.name = this.name || BlockUIDefaultName;
            this.template = this.template || this.blockUI.blockUISettings.template;
        };
        BlockUIComponent.ctorParameters = function () { return [
            { type: BlockUIInstanceService }
        ]; };
        __decorate([
            core.Input()
        ], BlockUIComponent.prototype, "name", void 0);
        __decorate([
            core.Input()
        ], BlockUIComponent.prototype, "message", void 0);
        __decorate([
            core.Input()
        ], BlockUIComponent.prototype, "delayStart", void 0);
        __decorate([
            core.Input()
        ], BlockUIComponent.prototype, "delayStop", void 0);
        __decorate([
            core.Input()
        ], BlockUIComponent.prototype, "template", void 0);
        BlockUIComponent = __decorate([
            core.Component({
                selector: 'block-ui',
                template: "\n    <ng-content></ng-content>\n    <block-ui-content\n      [name]=\"name\"\n      [message]=\"message\"\n      [template]=\"template\"\n      [delayStart]=\"delayStart\"\n      [delayStop]=\"delayStop\"\n    >\n    </block-ui-content>\n  ",
                encapsulation: core.ViewEncapsulation.None
            })
        ], BlockUIComponent);
        return BlockUIComponent;
    }());

    // Spinner style - https://github.com/lukehaas/css-loaders
    var styles = "\n.block-ui-wrapper {\n  display: none;\n  position: fixed;\n  height: 100%;\n  width: 100%;\n  top: 0;\n  left: 0;\n  background: rgba(0, 0, 0, 0.70);\n  z-index: 30000;\n  cursor: wait;\n}\n\n.block-ui-wrapper.block-ui-wrapper--element {\n  position: absolute;\n}\n\n.block-ui-wrapper.active {\n  display: block;\n}\n\n.block-ui-wrapper.block-ui-main {\n  position: fixed;\n}\n\n.block-ui-spinner,\n.block-ui-template {\n  position: absolute;\n  top: 40%;\n  margin: 0 auto;\n  left: 0;\n  right: 0;\n  transform: translateY(-50%);\n}\n\n.block-ui-spinner > .message {\n  font-size: 1.3em;\n  text-align: center;\n  color: #fff;\n}\n\n.block-ui__element {\n  position: relative;\n}\n\n.loader,\n.loader:after {\n  border-radius: 50%;\n  width: 10em;\n  height: 10em;\n}\n.loader {\n  margin: 7px auto;\n  font-size: 5px;\n  position: relative;\n  text-indent: -9999em;\n  border-top: 1.1em solid rgba(255, 255, 255, 0.2);\n  border-right: 1.1em solid rgba(255, 255, 255, 0.2);\n  border-bottom: 1.1em solid rgba(255, 255, 255, 0.2);\n  border-left: 1.1em solid #ffffff;\n  -webkit-transform: translateZ(0);\n  -ms-transform: translateZ(0);\n  transform: translateZ(0);\n  -webkit-animation: load8 1.1s infinite linear;\n  animation: load8 1.1s infinite linear;\n}\n\n@-webkit-keyframes load8 {\n  0% {\n    -webkit-transform: rotate(0deg);\n    transform: rotate(0deg);\n  }\n  100% {\n    -webkit-transform: rotate(360deg);\n    transform: rotate(360deg);\n  }\n}\n\n@keyframes load8 {\n  0% {\n    -webkit-transform: rotate(0deg);\n    transform: rotate(0deg);\n  }\n  100% {\n    -webkit-transform: rotate(360deg);\n    transform: rotate(360deg);\n  }\n}\n";

    var template = "\n<div class=\"block-ui-wrapper {{name}} {{className}}\" [ngClass]=\"{ 'active': state.blockCount > 0 }\">\n  <div class=\"block-ui-spinner\" *ngIf=\"!templateCmp\">\n    <div class=\"loader\"></div>\n    <div *ngIf=\"message || defaultMessage\" class=\"message\">\n      {{ message || defaultMessage }}\n    </div>\n  </div>\n  <ng-template *ngIf=\"templateCmp\" #templateOutlet></ng-template>\n</div>\n";

    var BlockUIContentComponent = /** @class */ (function () {
        function BlockUIContentComponent(blockUI, resolver, changeDetectionRef) {
            this.blockUI = blockUI;
            this.resolver = resolver;
            this.changeDetectionRef = changeDetectionRef;
            this.name = BlockUIDefaultName;
            this.defaultBlockState = {
                startTimeouts: [],
                stopTimeouts: [],
                updateTimeouts: [],
                blockCount: 0,
                startCallCount: 0,
                stopCallCount: 0
            };
            this.state = __assign({}, this.defaultBlockState);
        }
        BlockUIContentComponent.prototype.ngOnInit = function () {
            this.settings = this.blockUI.getSettings();
            this.blockUISubscription = this.subscribeToBlockUI(this.blockUI.observe());
        };
        BlockUIContentComponent.prototype.ngAfterViewInit = function () {
            try {
                if (!this.templateCmp) {
                    return false;
                }
                if (this.templateCmp instanceof core.TemplateRef) {
                    this.templateOutlet.createEmbeddedView(this.templateCmp);
                }
                else {
                    var templateComp = this.resolver.resolveComponentFactory(this.templateCmp);
                    this.templateCompRef = this.templateOutlet.createComponent(templateComp);
                    this.updateBlockTemplate(this.message);
                }
            }
            catch (error) {
                console.error('ng-block-ui:', error);
            }
        };
        BlockUIContentComponent.prototype.ngAfterViewChecked = function () {
            this.detectChanges();
        };
        BlockUIContentComponent.prototype.subscribeToBlockUI = function (blockUI$) {
            var _this = this;
            return blockUI$.subscribe(function (event) { return _this.onDispatchedEvent(event); });
        };
        BlockUIContentComponent.prototype.onDispatchedEvent = function (event) {
            switch (event.action) {
                case BlockUIActions.START:
                    this.onStart(event);
                    break;
                case BlockUIActions.STOP:
                    this.onStop(event);
                    break;
                case BlockUIActions.UPDATE:
                    this.onUpdate(event);
                    break;
                case BlockUIActions.RESET:
                    this.onReset(event);
                    break;
                case BlockUIActions.RESET_GLOBAL:
                    this.resetState();
                    break;
                case BlockUIActions.UNSUBSCRIBE:
                    this.onStop(event);
                    this.onUnsubscribe(event.name);
                    break;
            }
        };
        BlockUIContentComponent.prototype.onStart = function (_a) {
            var _this = this;
            var name = _a.name, message = _a.message;
            if (name === this.name) {
                var delay = this.delayStart || this.settings.delayStart || 0;
                this.state.startCallCount += 1;
                var startTimeout = setTimeout(function () {
                    _this.state.blockCount += 1;
                    _this.showBlock(message);
                    _this.updateInstanceBlockCount();
                }, delay);
                this.state.startTimeouts.push(startTimeout);
            }
        };
        BlockUIContentComponent.prototype.onStop = function (_a) {
            var _this = this;
            var name = _a.name;
            if (name === this.name) {
                var stopCount = this.state.stopCallCount + 1;
                if (this.state.startCallCount - stopCount >= 0) {
                    var delay = this.delayStop || this.settings.delayStop || 0;
                    this.state.stopCallCount = stopCount;
                    var stopTimeout = setTimeout(function () {
                        _this.state.blockCount -= 1;
                        _this.updateInstanceBlockCount();
                        _this.detectChanges();
                    }, delay);
                    this.state.stopTimeouts.push(stopTimeout);
                }
            }
        };
        BlockUIContentComponent.prototype.onUpdate = function (_a) {
            var _this = this;
            var name = _a.name, message = _a.message;
            if (name === this.name) {
                var delay = this.delayStart || this.settings.delayStart || 0;
                clearTimeout(this.state.updateTimeouts[0]);
                var updateTimeout = setTimeout(function () {
                    _this.updateMessage(message);
                }, delay);
                this.state.updateTimeouts.push(updateTimeout);
            }
        };
        BlockUIContentComponent.prototype.onReset = function (_a) {
            var name = _a.name;
            if (name === this.name) {
                this.resetState();
            }
        };
        BlockUIContentComponent.prototype.updateMessage = function (message) {
            this.showBlock(message);
        };
        BlockUIContentComponent.prototype.showBlock = function (message) {
            this.message = message || this.defaultMessage || this.settings.message;
            this.updateBlockTemplate(this.message);
            this.detectChanges();
        };
        BlockUIContentComponent.prototype.updateBlockTemplate = function (msg) {
            if (this.templateCompRef && this.templateCompRef instanceof core.ComponentRef) {
                this.templateCompRef.instance.message = msg;
            }
        };
        BlockUIContentComponent.prototype.resetState = function () {
            __spread(this.state.startTimeouts, this.state.stopTimeouts, this.state.updateTimeouts).forEach(clearTimeout);
            this.state = __assign({}, this.defaultBlockState);
            this.updateInstanceBlockCount();
            this.detectChanges();
        };
        BlockUIContentComponent.prototype.onUnsubscribe = function (name) {
            if (this.blockUISubscription && name === this.name) {
                this.blockUISubscription.unsubscribe();
            }
        };
        BlockUIContentComponent.prototype.updateInstanceBlockCount = function () {
            if (this.blockUI.blockUIInstances[this.name]) {
                var blockCount = this.state.blockCount;
                this.blockUI.blockUIInstances[this.name].blockCount = blockCount;
            }
        };
        BlockUIContentComponent.prototype.detectChanges = function () {
            if (!this.changeDetectionRef['destroyed']) {
                this.changeDetectionRef.detectChanges();
            }
        };
        BlockUIContentComponent.prototype.ngOnDestroy = function () {
            this.resetState();
            this.onUnsubscribe(this.name);
            this.blockUI.clearInstance(this.name);
        };
        BlockUIContentComponent.ctorParameters = function () { return [
            { type: BlockUIInstanceService },
            { type: core.ComponentFactoryResolver },
            { type: core.ChangeDetectorRef }
        ]; };
        __decorate([
            core.Input()
        ], BlockUIContentComponent.prototype, "name", void 0);
        __decorate([
            core.Input()
        ], BlockUIContentComponent.prototype, "delayStart", void 0);
        __decorate([
            core.Input()
        ], BlockUIContentComponent.prototype, "delayStop", void 0);
        __decorate([
            core.Input('message')
        ], BlockUIContentComponent.prototype, "defaultMessage", void 0);
        __decorate([
            core.Input('template')
        ], BlockUIContentComponent.prototype, "templateCmp", void 0);
        __decorate([
            core.ViewChild('templateOutlet', { read: core.ViewContainerRef })
        ], BlockUIContentComponent.prototype, "templateOutlet", void 0);
        BlockUIContentComponent = __decorate([
            core.Component({
                selector: 'block-ui-content',
                template: template,
                encapsulation: core.ViewEncapsulation.None,
                styles: [styles]
            })
        ], BlockUIContentComponent);
        return BlockUIContentComponent;
    }());

    var BlockUIService = /** @class */ (function () {
        function BlockUIService(blockUIInstance) {
            this.blockUIInstance = blockUIInstance;
            this.globalDispatch = this.blockUIInstance.decorate();
        }
        /**
        * Starts blocking for given BlockUI instance or instances
        */
        BlockUIService.prototype.start = function (target, message) {
            this.dispatch(target, BlockUIActions.START, message);
        };
        /**
        * Stops blocking for given BlockUI instance or instances
        */
        BlockUIService.prototype.stop = function (target) {
            this.dispatch(target, BlockUIActions.STOP);
        };
        /**
        * Reset blocking for given BlockUI instance or instances
        */
        BlockUIService.prototype.reset = function (target) {
            this.dispatch(target, BlockUIActions.RESET);
        };
        /**
        * Reset blocking for all BlockUI instances
        */
        BlockUIService.prototype.resetGlobal = function () {
            this.globalDispatch.resetGlobal();
        };
        /**
        * Updates message for given BlockUI instance or instances
        */
        BlockUIService.prototype.update = function (target, message) {
            this.dispatch(target, BlockUIActions.UPDATE, message);
        };
        /**
        * Unsubscribes for given BlockUI instance or instances
        */
        BlockUIService.prototype.unsubscribe = function (target) {
            this.dispatch(target, BlockUIActions.UNSUBSCRIBE);
        };
        /**
        * Checks if BlockUI is actively blocking
        */
        BlockUIService.prototype.isActive = function (target) {
            if (target === void 0) { target = null; }
            var targets = target ? this.toArray(target) : null;
            var instances = this.blockUIInstance.blockUIInstances;
            return Object.keys(instances).some(function (key) {
                if (!targets) {
                    return instances[key].isActive;
                }
                return targets.indexOf(instances[key].name) >= 0 && instances[key].isActive;
            });
        };
        BlockUIService.prototype.dispatch = function (target, type, message) {
            var _this = this;
            if (target === void 0) { target = []; }
            var instances = this.toArray(target);
            instances.forEach(function (i) { return _this.blockUIInstance.decorate(i)[type](message); });
        };
        BlockUIService.prototype.toArray = function (target) {
            if (target === void 0) { target = []; }
            return typeof target === 'string' ? [target] : target;
        };
        BlockUIService.ctorParameters = function () { return [
            { type: BlockUIInstanceService }
        ]; };
        BlockUIService = __decorate([
            core.Injectable()
        ], BlockUIService);
        return BlockUIService;
    }());

    var BlockUIDirective = /** @class */ (function () {
        function BlockUIDirective(blockUIService, blockUIInstanceService, viewRef, templateRef, renderer, componentFactoryResolver) {
            this.blockUIService = blockUIService;
            this.blockUIInstanceService = blockUIInstanceService;
            this.viewRef = viewRef;
            this.templateRef = templateRef;
            this.renderer = renderer;
            this.componentFactoryResolver = componentFactoryResolver;
        }
        Object.defineProperty(BlockUIDirective.prototype, "blockUI", {
            set: function (name) { this.blockTarget = name; },
            enumerable: true,
            configurable: true
        });
        ;
        Object.defineProperty(BlockUIDirective.prototype, "blockUIMessage", {
            set: function (message) { this.message = message; },
            enumerable: true,
            configurable: true
        });
        ;
        Object.defineProperty(BlockUIDirective.prototype, "blockUITemplate", {
            set: function (template) { this.template = template; },
            enumerable: true,
            configurable: true
        });
        ;
        Object.defineProperty(BlockUIDirective.prototype, "blockUIDelayStart", {
            set: function (delayStart) {
                this.delayStart = delayStart ? Number(delayStart) : null;
            },
            enumerable: true,
            configurable: true
        });
        ;
        Object.defineProperty(BlockUIDirective.prototype, "blockUIDelayStop", {
            set: function (delayStop) {
                this.delayStop = delayStop ? Number(delayStop) : null;
            },
            enumerable: true,
            configurable: true
        });
        ;
        BlockUIDirective.prototype.ngOnInit = function () {
            try {
                this.viewRef.createEmbeddedView(this.templateRef);
                var parentElement = this.getParentElement();
                if (parentElement && !this.isComponentInTemplate(parentElement)) {
                    this.renderer.addClass(parentElement, 'block-ui__element');
                    this.blockUIComponentRef = this.createComponent();
                    var blockUIContent = this.findContentNode(this.viewRef.element.nativeElement);
                    if (blockUIContent) {
                        var settings = this.blockUIInstanceService.getSettings();
                        parentElement.appendChild(blockUIContent);
                        this.blockUIComponentRef.instance.className = 'block-ui-wrapper--element';
                        this.blockUIComponentRef.instance.name = this.blockTarget || BlockUIDefaultName;
                        if (this.message)
                            this.blockUIComponentRef.instance.defaultMessage = this.message;
                        if (this.delayStart)
                            this.blockUIComponentRef.instance.delayStart = this.delayStart;
                        if (this.delayStop)
                            this.blockUIComponentRef.instance.delayStop = this.delayStop;
                        if (this.template || settings.template)
                            this.blockUIComponentRef.instance.templateCmp = this.template || settings.template;
                    }
                }
            }
            catch (error) {
                console.error('ng-block-ui:', error);
            }
        };
        BlockUIDirective.prototype.isComponentInTemplate = function (element) {
            // Needed because of https://github.com/microsoft/TypeScript/issues/26235
            var targetElement = element || {};
            var children = targetElement.children;
            children = Array.from(children || []).reverse();
            return children.some(function (el) { return el && el.localName === 'block-ui'; });
        };
        BlockUIDirective.prototype.getParentElement = function () {
            var embeddedView = this.viewRef.get(0);
            return embeddedView.rootNodes[0];
        };
        // Needed for IE (#17)
        BlockUIDirective.prototype.findContentNode = function (element) {
            var nextSibling = element.nextSibling || {};
            var previousSibling = element.previousSibling || {};
            return [
                nextSibling,
                nextSibling.nextSibling,
                previousSibling,
                previousSibling.previousSibling
            ].find(function (e) { return e && e.localName === 'block-ui-content'; });
        };
        BlockUIDirective.prototype.createComponent = function () {
            var resolvedBlockUIComponent = this.componentFactoryResolver.resolveComponentFactory(BlockUIContentComponent);
            return this.viewRef.createComponent(resolvedBlockUIComponent);
        };
        BlockUIDirective.prototype.ngOnDestroy = function () {
            if (this.blockTarget) {
                this.blockUIService.reset(this.blockTarget);
            }
        };
        BlockUIDirective.ctorParameters = function () { return [
            { type: BlockUIService },
            { type: BlockUIInstanceService },
            { type: core.ViewContainerRef },
            { type: core.TemplateRef },
            { type: core.Renderer2 },
            { type: core.ComponentFactoryResolver }
        ]; };
        __decorate([
            core.Input()
        ], BlockUIDirective.prototype, "blockUI", null);
        __decorate([
            core.Input()
        ], BlockUIDirective.prototype, "blockUIMessage", null);
        __decorate([
            core.Input()
        ], BlockUIDirective.prototype, "blockUITemplate", null);
        __decorate([
            core.Input()
        ], BlockUIDirective.prototype, "blockUIDelayStart", null);
        __decorate([
            core.Input()
        ], BlockUIDirective.prototype, "blockUIDelayStop", null);
        BlockUIDirective = __decorate([
            core.Directive({ selector: '[blockUI]' })
        ], BlockUIDirective);
        return BlockUIDirective;
    }());

    var BlockUIServiceInstance = new BlockUIInstanceService();
    // Needed for AOT compiling
    var BlockUIModuleSettings = new core.InjectionToken('BlockUIModuleSettings');
    function provideInstance(settings) {
        BlockUIServiceInstance.updateSettings(settings);
        return BlockUIServiceInstance;
    }
    var BlockUIModule = /** @class */ (function () {
        function BlockUIModule() {
        }
        BlockUIModule_1 = BlockUIModule;
        BlockUIModule.forRoot = function (settings) {
            if (settings === void 0) { settings = {}; }
            return {
                ngModule: BlockUIModule_1,
                providers: [
                    {
                        provide: BlockUIModuleSettings,
                        useValue: settings
                    },
                    {
                        provide: BlockUIInstanceService,
                        useFactory: provideInstance,
                        deps: [BlockUIModuleSettings]
                    },
                    BlockUIService
                ]
            };
        };
        var BlockUIModule_1;
        BlockUIModule = BlockUIModule_1 = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule
                ],
                entryComponents: [
                    BlockUIComponent,
                    BlockUIContentComponent
                ],
                declarations: [
                    BlockUIComponent,
                    BlockUIDirective,
                    BlockUIContentComponent
                ],
                exports: [
                    BlockUIComponent,
                    BlockUIDirective,
                    BlockUIContentComponent
                ]
            })
        ], BlockUIModule);
        return BlockUIModule;
    }());

    var blockInstanceGuid = 1;
    function BlockUI(blockName, settings) {
        if (settings === void 0) { settings = {}; }
        if (!settings.scopeToInstance) {
            return function (target, propertyKey) {
                target[propertyKey] = BlockUIServiceInstance.decorate(blockName);
            };
        }
        return function (target, key) {
            var secret = "_" + key + "-block-ui";
            Object.defineProperty(target, key, {
                get: function () {
                    if (this[secret]) {
                        return this[secret];
                    }
                    var instanceName = blockName + "-" + blockInstanceGuid++;
                    this[secret] = BlockUIServiceInstance.decorate(instanceName);
                    return this[secret];
                },
                set: function (value) {
                    this[secret] = value;
                },
            });
        };
    }

    exports.BLOCKUI_DEFAULT = BlockUIDefaultName;
    exports.BlockUI = BlockUI;
    exports.BlockUIComponent = BlockUIComponent;
    exports.BlockUIContentComponent = BlockUIContentComponent;
    exports.BlockUIModule = BlockUIModule;
    exports.BlockUIService = BlockUIService;
    exports.ɵa = BlockUIModuleSettings;
    exports.ɵb = provideInstance;
    exports.ɵc = BlockUIInstanceService;
    exports.ɵd = template;
    exports.ɵe = styles;
    exports.ɵf = BlockUIDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
